Initial implementation of the new container interface

alesgenova opened the pull request.

cjh1 committed:
Initial sketch of container interface

alesgenova committed:
Further define the schema of the container's self-describing obj

alesgenova committed:
Add a trivial cjson reader

alesgenova committed:
Add simple hashing function for dicts

alesgenova committed:
Adapt the python api to be able to run the new containers

alesgenova committed:
Implement a single taskflow that will work with all codes

alesgenova committed:
Upload jobs output manually

cjh1 commented on openchemistry/__init__.py:
```diff
@@ -348,17 +210,25 @@ def __init__(self, _id, cjson):
 
-    def optimize(self, basis=None, theory=None, functional=None, code='nwchem'):
-        return _optimize(self._id, basis, theory, functional, code=code)
+    def calculate(self, container_name, input_parameters, input_geometry=None):
```
@cryos How do you feel about this name?

cryos replied:
I was just starting to look at this, I don't think we should get too attached to it. I also think that even if we have an underlying generic call to execute_code or similar, that we will still want to have some high level calls that might just add type = optimize under the covers and dispatch to a very generic function.

This is part of what we were wanting to rework some, and I think it would benefit from some discussion.

cjh1 commented on openchemistry/__init__.py:
```diff
@@ -729,90 +598,6 @@ def products(self):
-        free_energy = free_energy * 2625.5
-
-        return free_energy
-
```
Lets not delete this, as its specific to NWChem, lets pull it out as an example. May be `openchemistry.examples.nwchem`?

cryos replied:
Agreed, we should move these to examples or similar. This is something you could do in many codes.

cjh1 commented on openchemistry/__init__.py:
```diff
@@ -885,48 +667,6 @@ def compose_equation(equation, **vars):
-    except ImportError:
-        # Outside notebook print the data
-        print(free_energy_chart_data)
-
```
Move to example

cryos replied:
:+1:

cjh1 commented on openchemistry/utils.py:
```diff
@@ -86,4 +87,7 @@ def calculate_mo(cjson, mo):
+    return json.loads(conv.write_string(mol, "cjson"))['cube']
+
+def hash_object(obj):
+    return hashlib.sha1(json.dumps(obj, sort_keys=True).encode()).hexdigest()
```
Lets use sha512

cryos replied:
:+1: :100: We talked about this yesterday evening, I said we should just use a sha512 if there was no good reason to avoid it. We need to break the habit of defaulting to sha1 in many places - there are far better algorithms and this code will be used in 2020 and beyond I hope!

cjh1 commented on taskflows/taskflows/__init__.py:
```diff
@@ -29,21 +30,26 @@ class OpenChemistryTaskFlow(TaskFlow, ABC):
             "_id": <id of cluster to run on>
-        }
+        },
+        "containerName": <The docker image name used to run the calculation>
```
imageName?

cjh1 replied:
We should split this out into `repositoryName` and `tag`

cjh1 commented on taskflows/taskflows/__init__.py:
```diff
@@ -155,196 +133,222 @@ def setup_input_template(task, input_, cluster):
+        submit_job(cluster, job, girder_token=task.taskflow.girder_token, monitor=False)
+    except:
+        import traceback
+        traceback.print_exc()
```
Let just remove this try except

cjh1 commented on taskflows/taskflows/__init__.py:
```diff
@@ -155,196 +133,222 @@ def setup_input_template(task, input_, cluster):
+
+    body = {
+        # ensure there are no special characters in the submission script name
+        'name': 'desc_%s' % re.sub('[^a-zA-Z0-9]', '_', container_name),
```
Can we create a PR for cumulus to raise an exception for the slash.

cryos commented:
So, high level, what does executing a job now look like? I feel like it may be a little more generic that we would want, but a few very minimal functions dispatching to the more generic function would return it to something..

cryos reviewed (commented):
So, high level, what does executing a job now look like? I feel like it may be a little more generic than we would want, but a few very minimal functions dispatching to the more generic function would return it to something similar to before with minimal maintenance overhead.

I think having a very generic entry point as suggested is nice, with a few convenience functions that really hard-wire optimization, energy, frequencies, etc would be nice to preserve.

alesgenova committed:
Update structure of the calculation model

alesgenova committed:
Address review comments

cjh1 committed:
Add utility script to pull images

alesgenova committed:
Update structure of the calculation model

alesgenova committed:
Merge pull request #47 from OpenChemistry/pull

Add utility script to pull images

alesgenova committed:
Pull the docker image through the pull.py script

alesgenova committed:
Fix typo

alesgenova committed:
Bring back the energy/optimize/frequencies methods

alesgenova committed:
Finish wiring up pull.py in the taskflow

alesgenova committed:
Add a way to pass options to the TaskFlow class

Used for now for things such as deciding whether to run with docker or
singularity, and to choose whether to keep or discard the scratch
folder. More options will be added in the future as needed

alesgenova committed:
Add ability to run singularity jobs in the dev demo
